File: wasm/lib/detailed/custom_jsons/rc.ts
import type { NaiAsset } from "../index";
import { HiveAppsOperation, TAccountName } from './factory.js';
import Long from 'long';

export class ResourceCreditsOperationData {
  public constructor(
    public readonly from: TAccountName,
    public readonly rc: NaiAsset,
    public readonly delegatees: Array<TAccountName>
  ) {}
}

export class ResourceCreditsOperation extends HiveAppsOperation<ResourceCreditsOperation> {
  protected readonly id = "rc";

  /**
   * Delegates resource credits to given account(s)
   *
   * @param {string} workingAccount The account which has to authorize transaction.
   *                                Also it will be the account which delegates resource credits (aka from)
   * @param {string | number | Long} maxRc maximum resource credits delegated (must pe non-negative value)
   * @param {TAccountName} delegatee target account to delegate. Remember you cannot delegate to yourself!
   * @param {TAccountName[]} otherDelegatees optional list of other target accounts to delegate.
   *                                         In the standard node configuration there may be 100 delegatees at most.
   *
   * @returns {ResourceCreditsOperation} itself
   */
  public delegate(workingAccount: TAccountName, maxRc: string | number | Long, delegatee: TAccountName, ...otherDelegatees: TAccountName[]): ResourceCreditsOperation {
    const delegatees = [ delegatee, ...otherDelegatees ];

    this.body.push([
      "delegate_rc",
      {
        from: workingAccount,
        delegatees,
        max_rc: Long.fromValue(maxRc),
        extensions: []
      }
    ]);

    return this;
  }

  /**
   * Removes resource credits delegation from given account(s)
   *
   * @param {string} workingAccount The account which has to authorize transaction.
   *                                Also it will be the account which delegated resource credits in the past (aka from)
   * @param {TAccountName} delegatee target account to remove the delegation from.
   * @param {TAccountName[]} otherDelegatees optional list of other target accounts to remove the delegation from.
   *
   * @returns {ResourceCreditsOperation} itself
   */
  public removeDelegation(workingAccount: TAccountName, delegatee: TAccountName, ...otherDelegatees: TAccountName[]): ResourceCreditsOperation {
    return this.delegate(workingAccount, 0, delegatee, ...otherDelegatees);
  }
}

File: python/wax/hive_apps_operations/rc.py
from dataclasses import dataclass
from typing import List

from wax.hive_apps_operations.factory import TAccountName, HiveAppsOperation, ChildT, BodyT
from wax.models.asset import NaiAsset, HiveNaiAssetConvertible

@dataclass
class ResourceCreditsOperationData:
    def __init__(self, from_: TAccountName, rc: NaiAsset, delegatees: List[TAccountName]):
        self.from_ = from_
        self.rc = rc
        self.delegatees = delegatees


class ResourceCreditsOperation(HiveAppsOperation[ChildT, BodyT]):
    @property
    def id(self) -> str:
        return "rc"

    def delegate(
        self,
        working_account: TAccountName,
        max_rc: HiveNaiAssetConvertible,
        delegatee: TAccountName,
        *other_delegatees: TAccountName
    ) -> 'ResourceCreditsOperation':
        delegatees = [delegatee, *other_delegatees]

        self.body.append([
            "delegate_rc",
            {
                "from": working_account,
                "delegatees": delegatees,
                "max_rc": str(max_rc),
                "extensions": []
            }
        ])

        return self

    def remove_delegation(
        self,
        working_account: TAccountName,
        delegatee: TAccountName,
        *other_delegatees: TAccountName
    ) -> 'ResourceCreditsOperation':
        return self.delegate(working_account, '0', delegatee, *other_delegatees)